### ISSUE TITLE
Xpra 4.1: SSH does not work on Windows 10 Client

### ISSUE THREAD

**Issue by shururuun** [Xpra-org/xpra#3040]
When connecting from a Xpra 4.1 Windows 10 Client (64 bit) to a Debian Server via SSH the Xpra client window gets the following error message:
```
_getfullpathname: path should be string, bytes or os.PathLike, not bool
```

**Comment by totaam** [comment]
Please provide the exact command line used. And if possible, run with "-d ssh".
What is the version on the server?

**Comment by tam116** [comment]
For me I'm using the GUI with everything at default. Connection type ssh. Server is 4.0.5 I tried using both hostname and IP but go the same error. Using the command line worked, C:\Program Files\Xpra\Xpra.exe' attach ssh://<myuser>@<myhost>/0

**Comment by shururuun** [comment]
Version on the server is: 4.1-r28561-1 amd65
I'm also using the GUI. The equivalent command line would be:

Xpra_cmd.exe attach ssh://user@172.23.255.250:22/23/

This generates the following error:
```
2021-03-04 08:22:47,708 Error: failed to receive anything, not an xpra server?
2021-03-04 08:22:47,710   could also be the wrong protocol, username, password or port
2021-03-04 08:22:47,711 remote SSH stderr:
2021-03-04 08:22:47,711   or the session was not found
2021-03-04 08:22:47,711  xpra main error:
2021-03-04 08:22:47,712 Connection lost
2021-03-04 08:22:47,713  Traceback (most recent call last):
2021-03-04 08:22:47,714    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 126, in main
2021-03-04 08:22:47,718      return run_mode(script_file, err, options, args, mode, defaults)
2021-03-04 08:22:47,720    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 400, in run_mode
2021-03-04 08:22:47,720      return do_run_mode(script_file, error_cb, options, args, mode, defaults)
2021-03-04 08:22:47,720    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 474, in do_run_mode
2021-03-04 08:22:47,720      return run_proxy(error_cb, options, script_file, args, mode, defaults)
2021-03-04 08:22:47,720    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 2751, in run_proxy
2021-03-04 08:22:47,721      display = pick_display(error_cb, opts, args)
2021-03-04 08:22:47,721    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 1126, in pick_display
2021-03-04 08:22:47,723      return parse_display_name(error_cb, opts, extra_args[0], session_name_lookup=True)
2021-03-04 08:22:47,724    File "/usr/lib/python3/dist-packages/xpra/scripts/main.py", line 803, in parse_display_name
2021-03-04 08:22:47,724      while display_name[pos]=="/" and scount<2:
2021-03-04 08:22:47,725  IndexError: string index out of range
```

Before that:
```
2021-03-04 08:22:44,294 Error: cannot enable SSH socket upgrades:
2021-03-04 08:22:44,295  The version of cryptography does not match the loaded shared object. This can happen if you have multiple copies of cryptography installed in your Python path. Please try creating a new virtual environment to resolve this issue. Loaded python version: 3.4.4, shared object version: b'2.9.2'
```

Adding "-d ssh" adds:
```
2021-03-04 08:28:33,270 import paramiko
Traceback (most recent call last):
  File "E:\Xpra/xpra/net/socket_util.py", line 331, in create_sockets
  File "<frozen zipimport>", line 259, in load_module
  File "C:/msys64/mingw64/lib/python3.8/site-packages/paramiko/__init__.py", line 22, in <module>
  File "<frozen zipimport>", line 259, in load_module
  File "C:/msys64/mingw64/lib/python3.8/site-packages/paramiko/transport.py", line 129, in <module>
  File "C:/msys64/mingw64/lib/python3.8/site-packages/paramiko/transport.py", line 190, in Transport
  File "C:/msys64/mingw64/lib/python3.8/site-packages/paramiko/kex_curve25519.py", line 30, in is_available
  File "C:/msys64/mingw64/lib/python3.8/site-packages/cryptography/hazmat/primitives/asymmetric/x25519.py", line 39, in generate
  File "C:/msys64/mingw64/lib/python3.8/site-packages/cryptography/hazmat/backends/openssl/__init__.py", line 6, in <module>
  File "C:/msys64/mingw64/lib/python3.8/site-packages/cryptography/hazmat/backends/openssl/backend.py", line 113, in <module>
  File "C:/msys64/mingw64/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 170, in <module>
  File "C:/msys64/mingw64/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line 159, in _verify_package_version
ImportError: The version of cryptography does not match the loaded shared object. This can happen if you have multiple copies of cryptography installed in your Python path. Please try creating a new virtual environment to resolve this issue. Loaded python version: 3.4.4, shared object version: b'2.9.2'
```

**Comment by totaam** [comment]
> `Xpra_cmd.exe attach ssh://user@172.23.255.250:22/23/`
> ..
> ` IndexError: string index out of range`

This was caused by the trailing slash in your URI. This works OK:
`Xpra_cmd.exe attach ssh://user@172.23.255.250:22/23`
From now on, we will ignore it: 111d3a6a159c57b8e7caa296eb4cc13a903b349e

---

The paramiko error:
> ```
> Error: cannot enable SSH socket upgrades:
> The version of cryptography does not match the loaded shared object...
> Loaded python version: 3.4.4, shared object version: b'2.9.2'
> ```

That's odd, I'm not seeing this here and the version numbers don't correspond to openssl or paramiko version numbers.
I have rebuilt the paramiko library, hopefully that will help. Please try the latest 4.2 beta build:
https://xpra.org/beta/windows/

---

As for the OP:
> `_getfullpathname: path should be string, bytes or os.PathLike, not bool`

@shururuun That's impossible for me to fix without further details.  
None of the command lines in this ticket produce this error and the `_getfullpathname` function is not one we call from xpra directly.

**Comment by shururuun** [comment]
I could track down the original error further with:

C:\Program Files\Xpra>Xpra-Launcher-Debug.exe -d ssh
2021-03-05 09:20:27,571 cannot connect:
Traceback (most recent call last):
  File "E:\Xpra/xpra/client/gtk_base/client_launcher.py", line 730, in do_connect
  File "E:\Xpra/xpra/client/gtk_base/client_launcher.py", line 773, in connect_builtin
  File "E:\Xpra/xpra/scripts/main.py", line 677, in add_ssh_args
  File "C:/msys64/mingw64/lib/python3.8/ntpath.py", line 540, in abspath
TypeError: _getfullpathname: path should be string, bytes or os.PathLike, not bool

![xpra-connect](https://user-images.githubusercontent.com/50909729/110088203-d65d4280-7d94-11eb-8286-b817daae0c69.gif)

**Comment by totaam** [comment]
@shururuun thanks for the details!
This bug should be fixed in 177058ebd38c6ff5a0e469986a2f3ad67e925752.

I was not seeing this problem because I am using the (default) paramiko ssh backend, but the build you used also had those paramiko library version problems you also reported (which should also be fixed?), so that's why you ended up taking a different path and hitting the bug.

The latest beta builds (r28593) has those fixes: https://xpra.org/beta/windows/

Please close if this works for you.

**Comment by shururuun** [comment]
Build revision 28593 fixed the SSH GUI problem for me. Thank you very much.
